clash-lib/src/Clash/Driver/BrokenGhcs.hs

```diff
@@ -0,0 +1,156 @@
+{-|
+Copyright   :  (C) 2024, Martijn Bastiaan
+License     :  BSD2 (see the file LICENSE)
+Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Utilities to detect and report GHC / operating system combinations that are
+known to be buggy.
+-}
+
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Driver.BrokenGhcs where
+
+import Data.Maybe (listToMaybe)
+import Data.Version (Version(Version, versionBranch))
+import GHC.Platform (OS(..))
+
+#if __GLASGOW_HASKELL__ > 810
+import System.Info (fullCompilerVersion)
+#endif
+
+import qualified Clash.Util.Interpolate as I
+import qualified System.Info
+
+#if __GLASGOW_HASKELL__ <= 810
+fullCompilerVersion :: Version
+fullCompilerVersion = System.Info.compilerVersion
+#endif
+
+-- | Current OS. Currently only recognizes Linux, Windows, and macOS.
+os :: OS
+os = case System.Info.os of
+  "darwin" -> OSDarwin
+  "linux" -> OSLinux
+  "mingw32" -> OSMinGW32
+  _ -> OSUnknown
+
+-- | What OS GHC is broken on (or all)
+data BrokenOn = All | SomeOs OS
+
+data GhcVersion
+  = Ghc Int Int Int
+  -- ^ @A.B.C@, for example @9.10.2@
+  deriving (Eq, Ord)
+
+data GhcRange = GhcRange
+  { from :: GhcVersion
+  -- ^ Start of range, inclusive
+  , to :: GhcVersion
+  -- ^ End of range, exclusive
+  }
+
+-- | Check if a 'GhcVersion' is within a 'GhcRange'
+ghcInRange :: GhcVersion -> GhcRange -> Bool
+ghcInRange ghc GhcRange{from, to} = from <= ghc && ghc < to
+
+-- | Construct a range of all GHC versions matching a major version
+ghcMajor :: Int -> Int -> GhcRange
+ghcMajor major minor = GhcRange{from=Ghc major minor 0, to=Ghc (major + 1) 0 0}
+
+data Why = Why
+  { what :: String
+    -- ^ What is broken
+  , solution :: String
+    -- ^ What can be done to work around or solve the issue
+  , issue :: String
+    -- ^ Link to issue
+  , brokenOn :: [(BrokenOn, GhcRange)]
+    -- ^ What operation systems are affected
+  }
+
+-- | Get current GHC version expressed as a triple. It probably does something
+-- non-sensible on unreleased GHCs.
+ghcVersion :: GhcVersion
+ghcVersion = Ghc major minor patch
+ where
+  (major, minor, patch) =
+    case fullCompilerVersion of
+      Version{versionBranch} ->
+        case versionBranch of
+          [] -> (0, 0, 1)
+          [a] -> (a, 0, 1)
+          [a, b] -> (a, b, 1)
+          [a, b, c] -> (a, b, c)
+          (a:b:c:_) -> (a, b, c)
+
+-- | Pretty print 'Why' into an error message
+whyPp :: Why -> String
+whyPp Why{what, solution, issue}= [I.i|
+  Clash has known issues on #{major}.#{minor}.#{patch} on your current
+  OS. While not completely preventing the compiler from working, we recommend
+  switching to another GHC version. Symptoms:
+
+    #{what}
+
+  Consider the following work around or solution:
+
+    #{solution}
+
+  More information can be found at:
+
+    #{issue}
+
+  If you want to ignore this message, pass the following flag to Clash:
+
+    -fclash-ignore-broken-ghcs
```

Comment:
This looks fine in code: a single blank line before the indented part and a single one after. But I tried it and there is a stray blank line before `#{what}`:
```
<no location info>: error:
    Clash error call:
    Clash has known issues on 9.4.8 on your current OS. While not completely preventing 
    the compiler from working, we recommend switching to another GHC version. Symptoms:
    
    
      Clash type error messages are indecipherable
    
    Consider the following work around or solution:
    
      Upgrade to GHC 9.4 or downgrade to GHC 9.0
    
    More information can be found at:
    
      <no link>
    
    If you want to ignore this message, pass the following flag to Clash:
    
      -fclash-ignore-broken-ghcs
    CallStack (from HasCallStack):
      error, called at src/Clash/Driver/BrokenGhcs.hs:156:15 in clash-lib-1.9.0-inplace:Clash.Driver.BrokenGhcs
```
(this also suffers from the `ghcMajor` function problem, but that's beside the point here)

My suspicion? When a paragraph is more than a single line of text, an extra newline gets tacked at the end of the reflowed paragraph. As that is a clear difference between the first paragraph and the subsequent ones which don't suffer from the issue.

Reply:
Okay, I've created an issue as this seems out of scope for this PR: https://github.com/clash-lang/clash-compiler/issues/2753. (Though it's probably an easy fix.)